### DISCUSSION TITLE
usb as HID

### DISCUSSION THREAD

**Discussion by kleffa** [darthcloud/BlueRetro#380] [Ideas]
There is a way to use blueretro as a usb receiver ?
Explained like to connect Bluetooth controllers as a PS3 controller conected with the usb cable.

I want to connect it to a Wii with fakemote, a mood to use PS3 controller by usb port, I guess there are others applies of this, so it could be interesting

**Comment by NickGravley** [comment]
Without combining code from the two projects, you may still be able to just use a PS3 console adapter and match the blueretro for the other side of the adapter. For example, you could build blueretro for Playstation 1/2 console, but then connect the blueretro to the Wii with a separate Dualshock 2 to Playstation 3 compatible USB adapter.  I have not tested this on wii+fakemote.  I use a similar setup for blueretro and original Xbox.

**Reply by darthcloud** [reply]
This is you best option.

**Reply by kleffa** [reply]
I know, but this loss the analog R2 and L3 or other functions, like tilt controller, also PS3 and xbox360 are able to accept usb controllers and there are several proyects that uses Arduino and ESP32 as keyboards o mouse's

**Comment by kleffa** [comment]
After trying some adapters (ps2-xbox360, ps2-ps3) those are not the best since lost of home buttons (I guest I can activate now with HW2) and also lost of some info like acelerometers and that stuff....

Some Arduino, like Leonardo or micro (At mega 32U4 based) have configurables USB modules that can be  configured as a joypad, but since CH340K and similar USB modules doesn't make this maybe putting a esp32 and a Arduino micro working together should be a great idea. 
Now the big question there is some way to comunicate bluretro to a Arduino? This can keep info of sensors and that to increase bluretro capabilities over usb?

**Reply by NickGravley** [reply]
Yes, you can still use the Playstation SPI for its support of many analogs. There are already some Playstation controller libraries that you can use to connect ESP32 to the Arduino, then on the PC/USB side you can use joyhid or xinput libraries.  Even though SPI can sometime run without level shifting, you may avoid potential problems if you plan...  You can use a 3.3v Arduino but these run at 8mhz usually, and not many support USB HID without the DFU hacks. Pro Micro are 5v/16MHz and are a cheap and easy joy hid, but you may require a level shifter.  Teensy-LC is 3.3v and reportedly supports many of the Arduino libraries including the xinput one.

**Reply by kleffa** [reply]
> Yes, you can still use the Playstation SPI for its support of many analogs. There are already some Playstation controller libraries that you can use to connect ESP32 to the Arduino, then on the PC/USB side you can use joyhid or xinput libraries. Even though SPI can sometime run without level shifting, you may avoid potential problems if you plan... You can use a 3.3v Arduino but these run at 8mhz usually, and not many support USB HID without the DFU hacks. Pro Micro are 5v/16MHz and are a cheap and easy joy hid, but you may require a level shifter. Teensy-LC is 3.3v and reportedly supports many of the Arduino libraries including the xinput one.

I guess it is kind of easy now this part, even bridging some GPIO in Arduino/teensy to chose xinput/hid or any other usb mode
Yes, it can receive esp32 as a PS2 joypad, but maybe there is a better way, to save sensor information as well, since this idea can become a extension for bluretro to extend it to Xbox/xbox360/ps3(this include Wii moods) and maybe newest consoles as well

What about a raspberry pi pico?

**Reply by NickGravley** [reply]
There is a project called GP2040 on pi pico.  I have not had a reason to try it yet, but may learn how for Switch.  I already have a pi zero collecting my usb game devices data and outputting i2c to avr receivers for wired mode and blegamepad to Blue Retro receivers for wireless setups.  I don't know what sensors you mention but would like to point out that serial is fast enough for some things and you could enable and filter serial data from blueretro without having to learn very much about its code.

**Reply by kleffa** [reply]
I made some code for a nano,to experiment this idea.. a ps2-(xbox360/PS3/switch/android) usb adapter....
With sensor I refer til control from Wiimote, DS3, joycon... Etc DS4 touch, microphone and those stuff if bluretro can share raw info via serial with the adapter it could be interpreted by the console

**Reply by kleffa** [reply]
After a while waiting for a nano that was lost on mail, I was thinking, how about a esp32-S3 since it have usb configurable interface like  AT32u4, there are several libraries for emulate a switch or PS3 controller, sadly Xbox have controller protections to avoid this
